web/api/tests/valid_urls.c

```diff
@@ -0,0 +1,777 @@
+{
+    (void)state;
+
+    if (localhost != NULL)
+        free(localhost);
+    localhost = malloc(sizeof(RRDHOST));
+
+    struct web_client *w = setup_fresh_web_client();
+    buffer_strcat(w->response.data, "GET http://localhost:19999/api/v1/info HTTP/1.1\r\n\r\n");
+
+    char debug[4096];
+    repr(debug, sizeof(debug), w->response.data->buffer, w->response.data->len);
+    printf("-> \"%s\"\n", debug);
+
+    //char expected_url_repr[4096];
+    //repr(expected_url_repr, sizeof(expected_url_repr), def->url_out_repr, strlen(def->url_out_repr));
+
+    expect_value(__wrap_web_client_api_request_v1, host, localhost);
+    expect_value(__wrap_web_client_api_request_v1, w, w);
+    expect_string(__wrap_web_client_api_request_v1, url_repr, "info");
+
+    web_client_process_request(w);
+
+    assert_string_equal(w->decoded_query_string, "?blah");
+    destroy_web_client(w);
+    free(localhost);
+    localhost = NULL;
+}
+
+static void valid_url(void **state)
+{
+    (void)state;
+
+    if (localhost != NULL)
+        free(localhost);
+    localhost = malloc(sizeof(RRDHOST));
+
+    struct web_client *w = setup_fresh_web_client();
+    buffer_strcat(w->response.data, "GET /api/v1/info?blah HTTP/1.1\r\n\r\n");
+
+    char debug[4096];
+    repr(debug, sizeof(debug), w->response.data->buffer, w->response.data->len);
+    printf("-> \"%s\"\n", debug);
+
+    //char expected_url_repr[4096];
+    //repr(expected_url_repr, sizeof(expected_url_repr), def->url_out_repr, strlen(def->url_out_repr));
+
+    expect_value(__wrap_web_client_api_request_v1, host, localhost);
+    expect_value(__wrap_web_client_api_request_v1, w, w);
+    expect_string(__wrap_web_client_api_request_v1, url_repr, "info");
+
+    web_client_process_request(w);
+
+    assert_string_equal(w->decoded_query_string, "?blah");
+    destroy_web_client(w);
+    free(localhost);
+    localhost = NULL;
+}
+
+/* RFC2616, section 4.1:
+
+   In the interest of robustness, servers SHOULD ignore any empty
+   line(s) received where a Request-Line is expected. In other words, if
+   the server is reading the protocol stream at the beginning of a
+   message and receives a CRLF first, it should ignore the CRLF.
+*/
+static void leading_blanks(void **state)
+{
+    (void)state;
+
+    if (localhost != NULL)
+        free(localhost);
+    localhost = malloc(sizeof(RRDHOST));
+
+    struct web_client *w = setup_fresh_web_client();
+    buffer_strcat(w->response.data, "\r\n\r\nGET /api/v1/info?blah HTTP/1.1\r\n\r\n");
+
+    char debug[4096];
+    repr(debug, sizeof(debug), w->response.data->buffer, w->response.data->len);
+    printf("-> \"%s\"\n", debug);
+
+    //char expected_url_repr[4096];
+    //repr(expected_url_repr, sizeof(expected_url_repr), def->url_out_repr, strlen(def->url_out_repr));
+
+    expect_value(__wrap_web_client_api_request_v1, host, localhost);
+    expect_value(__wrap_web_client_api_request_v1, w, w);
+    expect_string(__wrap_web_client_api_request_v1, url_repr, "info");
```

Comment:
url_repr -> url

Reply:
> Generally, you shouldn't have the "No entries for symbol" errors in the CMocka output.

This is not correct. Each of those cases is a case that fails because the tested code does the wrong thing. It should not call the API despatch point because it should not consider the URL to be valid (in each case it is malformed). The unit testing code will work correct (by asserting the return code) when the tested code is fixed.

I thought it would be cleaner to have the opposite of an `expect_` in the test to indicate that we should not hit the mock during the test but I could not see a way to do that.

Reply:
I don't understand it. `__wrap_web_client_api_request_v1, url_repr` doesn't have the `url_repr` parameter. How your `expect_` could help in identifying wrong calls? It fails because of the parameter name mismatch.

Reply:
The `expect_` macros are not limited to parameters, they can be applied to arbitrary values in the procedure being tested. In this case there is an exact 1:1 mapping from `url` to `url_repr`, but the `repr()` output is easier to work with in cases where the strings contain null characters.